include/wil/cppwinrt_notifiable_module_lock.h

```diff
@@ -0,0 +1,103 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+//! @file
+//! Utilities for implementing OOP COM server with cppwinrt only
+
+#ifndef __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_DEFINED
+#define __WIL_CPPWINRT_NOTIFIABLE_MODULE_LOCK_DEFINED
+
+#if _MSVC_LANG >= 201703L
+
+#include <atomic>
+#include <cstdint>
+
+// Ligh up cppwinrt custom module lock
+#define WINRT_CUSTOM_MODULE_LOCK
+
+namespace wil
+{
+// Adopted from cppwinrt
+template <typename Func>

```

Comment:
Comment now that I see your test - I don't think much value is added by allowing functions with a non-`void` return type. Simplifying here seems much better overall.

Reply:
This is just following the footstep of `WRL`'s, `Module`, which allows the callback to be generic. I already omitted the class method variant. I personally don't mind forcing the callback to be just `void callback()`, but I'm not sure if that would break any use case. Might need to ping other WRL users for insight here 🙏

Reply:
I think I can reference how wil's `scoped_exit` handle this case.

Reply:
Even `scope_exit` uses template🧐I bet the intention is to allow efficient storage of all callable - is `scope_exit` is avoiding std::function intentionally?

Reply:
scope_exit is more performance sensitive, it is used as a cheap wrapper for RAII where no other suitable one exists. As such it can show up in tight loops and I would imagine ends up inlined most of the time. Allocating a function and preventing inlining via type erasure could be a problem.

This is in contrast to our server here, which allocates a std::function once at startup (which is already an expensive operation) and then only calls it during shutdown. Not very performance sensitive.

Arguably, there are no concerns about code size bloat since you should probably only have 1 server in a single exe, but maintainability is a concern and templates generally result in suboptimal IDE experience.

Reply:
I agree with everything @sylveon says but will add one more thing: complexity. Not just the complexity of definition & maintenance, but the complexity in use. If I wanted to use something like a function object (which cannot be casted to a function pointer), I would need to dig around to figure out what additional compile definitions I need, what needs to be defined, etc. This is made even worse if I want to use a stateful lambda (with a non-empty capture) which not only also cannot be casted to a function pointer, but I also now have the additional complexity of trying to reference the lambda type from inside of my `get_module_lock` function. I could, of course, just use `notifiable_module_lock<std::function<void()>>`, but I see no reason we shouldn't just make this the default experience.

Reply:
> if I want to use a stateful lambda (with a non-empty capture) which not only also cannot be casted to a function pointer, but I also now have the additional complexity of trying to reference the lambda type from inside of my get_module_lock function

Technically you can go `decltype(myLambda)` but I agree it's not a very pleasant API. WRL go over the board in doing placement new and allocating the function storage for the generic type to hide the generic specification - I don't think we need that here either. Would go with `std::function` then.